### PR TITLE
[MRG] add boostrapped confidence intervals for ERPs

### PR DESCRIPTION
**Pull request by JoseAlanis**
Add proposal on how to fir a boostrapped linear model to full sensor data.
Quick summary:

- Data is shuffled of resampled a particular number of times (i.e., variable boot). 
- Linear model is fitted to all of the resampled data sets. Result is stored for further manipulation.
- Betas for all models are extracted and a centrality estimator is computed (currently the mean).
- The mean beta coefficient is plotted in a "EEG-fashion".

### CODE REVIEW

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    beta = betas[:, ind]
+    # back projection to channels x time points
+    beta = beta.reshape((n_channels, n_times))
+    # create evoked object containing the back projected coefficients
+    # for each predictor
+    lm_betas[predictor] = EvokedArray(beta, epochs_info, tmin)
+
+
+###############################################################################
+# -- initialize bootstrap procedure --
```
Comment (JoseAlanis) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: this is where it gets interesting

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+# shape of the resulting data (number of random samples x shape of Y)
+boot_shape = (boot, n_epochs, n_channels * n_times)
+
+# create sets of random samples
+resamples = []
+for i in range(boot):
+    resamples.append(list(np.random.choice(range(n_epochs), n_epochs, replace=True)))  # noqa
+
+# fit linear model to each of the random samples of Y
+boot_reg = [boot_linear_model.fit(design.iloc[resamples[i]], Y[resamples[i], :]) for i in range(boot)]  # noqa
```
Comment (JoseAlanis) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: I was thinking whether it would be important to "rescale" predictors each time we fit the model? Would not doing this introduce unnecessary random variation to the model?

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+# extract coefficients
+boot_betas = [get_coef(reg, 'coef_') for reg in boot_reg]
+
+# create a centrality estimator (here the mean) for the linear models fitted
+# on the random samples.
+centrality_estimator = np.mean(np.asarray(boot_betas), axis=0)
```
Comment (JoseAlanis) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: I'm currently taking the mean of all betas, but it might be interesting to allow users the estimator the want to use ... does anyone use median ERPs?

Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: One option would be to return the individual Evoked instances and use mne.combine_evoked, which has options, for the averaging.

Comment (JoseAlanis) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: actually, haven't thought of that, but we could use the weights argument to apply [Dirichlet weights](https://en.wikipedia.org/wiki/Dirichlet_distribution) to the data. This would go in the direction of the Bayesian bootstrap used in the LIMO toolbox (see [here](https://github.com/LIMO-EEG-Toolbox/limo_eeg/blob/master/limo_central_estimator.m#L52-L54))

Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: Well we can leave the door open to that, but maybe start with a mean option and stay modular?

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+================================================================
+
+"""
+
+# Authors: Jose C. Garcia Alanis <alanis.jcg@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+from sklearn.linear_model import LinearRegression
```
Comment (jona-sassenhagen) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: Can you maybe time this vs. Ridge (with alpha=0)?

Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: I think the backbone is different, and that Ridge with Cholesky should be fastest for multiple targets.

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+# pick channels that should be included in the analysis
+picks_eeg = pick_types(limo_epochs['2'].info, eeg=True)
+# channels to be excluded
+exclude = ['EXG1', 'EXG2', 'EXG3', 'EXG4']
+
+# save epochs information (needed for creating a homologous
+# epochs object containing linear regression result)
+epochs_info = limo_epochs['2'].copy().drop_channels(exclude).info
+tmin = limo_epochs['2'].tmin.copy()
```
Comment (jona-sassenhagen) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: Why copy an int?

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 2 above)
Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: Just break this line and remove the noqa

Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: Also you should probably loop over resamples hm? Not over range.

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 3 above)
Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: Also if you're using mean, use the method not the function :)

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+# visualise effect of phase-coherence for sklearn estimation method.
+lm_betas['phase-coherence'].plot_joint(ts_args=ts_args,
+                                       title='Phase-coherence (sklearn betas)',
+                                       times=[.23])
+
+###############################################################################
+# visualise effect of phase-coherence for bootstrapped  method.
+boot_lm_betas['phase-coherence'].plot_joint(ts_args=ts_args,
+                                            title='Phase-coherence (boot betas)',  # noqa
```
Comment (jona-sassenhagen) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: Just break properly, as by PEP8. I.e., 

```
boot_lm_betas['phase-coherence'].plot_joint(
    ts_args=ts_args, title=...
```

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+# number of random samples
+boot = 100
+# shape of the resulting data (number of random samples x shape of Y)
+boot_shape = (boot, n_epochs, n_channels * n_times)
+
+# create sets of random samples
+resamples = []
+for i in range(boot):
+    resamples.append(list(np.random.choice(range(n_epochs), n_epochs, replace=True)))  # noqa
```
Comment (jona-sassenhagen) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: Why are you using a loop here and a list comp in 111?

File: examples/plot_bootstrapped_coffecients_for_lm.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+linear_model = LinearRegression(fit_intercept=False)
+linear_model.fit(design, Y)
+
+# extract the resulting coefficients (i.e., betas)
+betas = get_coef(linear_model, 'coef_')
+
+# loop through the columns (i.e., the predictors) of coefficient matrix
+# and extract coefficients for each predictor.
+lm_betas = dict()
+for ind, predictor in enumerate(predictors):
```
Comment (jona-sassenhagen) [inline examples/plot_bootstrapped_coffecients_for_lm.py]: Or zip over predictors and betas.T

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 2 above)
Comment (jona-sassenhagen) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: Not sure I understand what you mean.

**Comment by JoseAlanis** [comment]
the rendered output of the bootstrapped regression
boot_betas

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 5 above)
Comment (JoseAlanis) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: this is just me being too anxious about shallow copies :grimacing:

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 8 above)
Comment (JoseAlanis) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: That was actually the way we did it before, I just though it would better to keep the shape of the matrices (for consistence) and then just loop through the columns.

**Comment by dengemann** [comment]
Ok we need fo talk. This is not yet the right output. I’d like to see confidence intervals in a topo plot ideally. It is not clear to me what tis output is showing.

**Comment by jona-sassenhagen** [comment]
With a linear model, just averaging the betas from the bootstrap runs will converge on the least-square solution ...

**Comment by dengemann** [comment]
Yes taking the average of a bootstrap distro is not directly of interest and not its intended use.

**Comment by JoseAlanis** [comment]
bootstrap ci example for electrode B29
face_a_erp_ci

**Comment by JoseAlanis** [comment]
Alright, I updated the code according to your comments. The bootstrap does look better and easier to understand in one for loop.

**Comment by dengemann** [comment]
What is the reason to have 2 examples here? I see 2 files.

**Comment by JoseAlanis** [comment]
The file `plot_bootstrapped_coffecients_for_lm.py` contains the code I used to bootstrap the coefficients from `sklearn.linear_model.LinearRegression`. I've put it on ice after noticing the approach was going in the wrong direction and decided to work first on the bootstrap of the condition ERPs. I'll remove the `plot_bootstrapped_coffecients_for_lm.py` file from this PR and open a new one for further discussion.

**Comment by jona-sassenhagen** [comment]
Sounds good!

File: examples/plot_bootstrapped_coffecients_for_lm.py (hunk 2 above)
Comment (JoseAlanis) [reply examples/plot_bootstrapped_coffecients_for_lm.py]: This is now part of another PR (see #6)

**Comment by JoseAlanis** [comment]
I updated the code according to the comments in #6. 
Ready to merge if everything ok.

File: examples/plot_bootstrap_ci.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+# number of random samples
+boot = 2000
+
+# initialize bootstrap
+resampled_data = []
+for i in range(boot):
+    # pick random samples with replacement from data
+    resamples = random.choice(range(n_epochs), n_epochs, replace=True)
+    # compute centrality estimator on re-sampled data
+    resampled_data.append(Y[resamples, :].mean(axis=0))
```
Comment (dengemann) [inline examples/plot_bootstrap_ci.py]: I think it would be more pedagogical to show how to bootstrap the fit, even if this may give identical results in this case.

Comment (JoseAlanis) [reply examples/plot_bootstrap_ci.py]: Ok, so then perhaps re-open #6 and close this one?

Comment (dengemann) [reply examples/plot_bootstrap_ci.py]: I think we/me got confused with all these prs. Well or copy over the few lines meded.

**Comment by JoseAlanis** [comment]
Ok, took care of the adjustments and also added some changes suggested by @jona-sassenhagen in #8. Should be ready to merge.

**Comment by JoseAlanis** [comment]
Hey guys, thanks for the suggestions and comments. I'll merge this PR for now. I think we've addressed the main issues and have a good basis. We can continue improving this section later.
